content/en/blog/_posts/2023-09-15-Bootstrap-an-Air-Gapped-Cluster-with-Kubeadm/index.md

```diff
@@ -0,0 +1,665 @@
+
+
+```bash
+mkdir -p /opt/cni/bin
+tar -C /opt/cni/bin -xz -f "cni-plugins-linux-${K8s_ARCH}-v1.3.0.tgz"
+tar -C /usr/local/bin-xz -f "crictl-v1.27.0-linux-${K8s_ARCH}.tar.gz"
+```
+
+**Make kubeadm, kubelet and kubectl executable and move them from the /tmp directory to /usr/local/bin:**
+```bash
+chmod +x kubeadm kubelet kubectl
+mv kubeadm kubelet kubectl /usr/local/bin
+```
+**Define an override for the systemd kubelet service file and move it to the proper location.**
+
+```bash
+mkdir -p /etc/systemd/system/kubelet.service.d
+
+sed "s:/usr/bin:/usr/local/bin:g" 10-kubeadm.conf > /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
+```
+
+**The CRI plugin for containerd is disabled by default; enable it:**
+```bash
+sed -i 's/^disabled_plugins = \["cri"\]/#&/' /etc/containerd/config.toml
+```
+**Put the /etc/docker/daemon.json file in place:**
+```bash
+echo '{
+
+"exec-opts": ["native.cgroupdriver=systemd"],
+
+"insecure-registries" : ["localhost:5000"],
+
+"allow-nondistributable-artifacts": ["localhost:5000"],
+
+"log-driver": "json-file",
+
+"log-opts": {
+
+"max-size": "100m"
+
+},
+
+"group": "rnd",
+
+"storage-driver": "overlay2",
+
+"storage-opts": [
+
+"overlay2.override_kernel_check=true"
+
+]
+
+}' > /etc/docker/daemon.json
+```
+Two important lines to highlight in the daemon.json file. The insecure-registries line means that the registry in brackets does not support TLS. Even inside an air gapped environment, this isn't a good practice but is fine for the purposes of this lab. The allow-nondistributable-artifacts line tells Docker to permit pushing nondistributable artifacts to this registry. Docker by default does not push these layers to avoid potential issues around licensing or distribution rights. A good example of this is the Windows base container image. This line will allow layers that Docker marks as "foreign" to be pushed to the registry. While not a big deal for this article, that line could be required for some air gapped environments. All layers have to exist locally since nothing inside the air gapped environment can reach out to a public container image registry to get what it needs.
+
+**Start and enable Docker:**
+
+```bash
+systemctl restart docker
+systemctl enable docker
+```
+
+**Start and enable the kubelet:**
+
+```bash
+systemctl start kubelet
+systemctl enable kubelet
+```
+
+The container image registry that runs in Docker is only required for any CNI related containers and subsequent workload containers. This registry is **not** used to house the Kubernetes component containers. Note, nerdctl would have also worked here as an alternative to Docker and would have allowed for direct interaction with containerd. Docker was chosen for its familiarity.
+
+**Start Docker registry**
+
+```bash
+docker load -i registry_2.8.2.tar
+docker run -d -p 5000:5000 --restart=always --name registry registry:2.8.2
+```
+
+**Load Flannel containers into the Docker registry:**
```

Comment:
we stopped recommending flannel some time ago, due to sporadic bugs and not keeping up to date with latest k8s.

you can check colium or kindnet for this demo. add note for users to pick their cni.

Reply:
This isn't the Kubernetes project recommending Flannel; this is @TheFutonEng picking a particular network plugin for a demo / lab.

@TheFutonEng you might like to outline - _briefly_ - how or why you picked this plugin. If the reasoning isn't important, you could make it clear that people reading this are free to pick a different plugin.

Reply:
@neolit123 , @sftim , thanks for the input.  The reasoning wasn't important behind the choice of Flannel.  I'll add some context here and make it clear that folks should choose the CNI that works best for their environment.